src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -884,15 +963,29 @@ impl<'tcx> Constructor<'tcx> {
                 ty::Slice(_) | ty::Array(..) => bug!("bad slice pattern {:?} {:?}", self, ty),
                 _ => PatKind::Wild,
             },
-            FixedLenSlice(_) => {
-                PatKind::Slice { prefix: subpatterns.collect(), slice: None, suffix: vec![] }
-            }
-            &VarLenSlice(prefix_len, _) => {
-                let prefix = subpatterns.by_ref().take(prefix_len as usize).collect();
-                let suffix = subpatterns.collect();
-                let wild = Pat::wildcard_from_ty(ty);
-                PatKind::Slice { prefix, slice: Some(wild), suffix }
-            }
+            Slice(slice) => match slice.pattern_kind() {
+                FixedLen(_) => {
+                    PatKind::Slice { prefix: subpatterns.collect(), slice: None, suffix: vec![] }
+                }
+                VarLen(prefix, _) => {
+                    let mut prefix: Vec<_> = subpatterns.by_ref().take(prefix as usize).collect();
+                    let mut suffix: Vec<_> = subpatterns.collect();
+                    if slice.array_len.is_some() {
+                        // Improves diagnostics a bit: if the type is a known-size array, instead
+                        // of reporting `[x, _, .., _, y]`, we prefer to report `[x, .., y]`.
+                        // This is incorrect if the size is not known, since `[_, ..]` captures
+                        // arrays of lengths `>= 1` whereas `[..]` captures any length.
+                        while !suffix.is_empty() && suffix.first().unwrap().is_wildcard() {
```

Comment:
Continually removing from the front of the vector seems like it could regress perf (tho the number of elements is likely small...)? Perhaps amortize the cost by first calculating the number of elements from the front to remove and then using `.drain(...)`. Alternatively, use `.skip_while(...)` when computing `suffix` (with a fast path for `slice.array_len.is_none()`) -- conversely, you can use `.take_while(...)` for `prefix` but that will need to go over everything until the first wildcard.